Source/WebCore/html/CanvasBase.cpp

```diff
@@ -264,39 +255,31 @@ void CanvasBase::setSize(const IntSize& size)
 
 RefPtr<ImageBuffer> CanvasBase::setImageBuffer(RefPtr<ImageBuffer>&& buffer) const
 {
-    RefPtr<ImageBuffer> returnBuffer;
-    {
-        Locker locker { m_imageBufferAssignmentLock };
-        m_contextStateSaver = nullptr;
-        returnBuffer = std::exchange(m_imageBuffer, WTFMove(buffer));
-    }
-
-    auto* context = renderingContext();
-
-    if (m_imageBuffer && m_size != m_imageBuffer->truncatedLogicalSize()) {
-        m_size = m_imageBuffer->truncatedLogicalSize();
-
-        if (context)
-            InspectorInstrumentation::didChangeCanvasSize(*context);
-    }
-
-    size_t previousMemoryCost = m_imageBufferCost;
-    m_imageBufferCost = memoryCost();
-    s_activePixelMemory += m_imageBufferCost - previousMemoryCost;
-
-    if (context && m_imageBuffer && previousMemoryCost != m_imageBufferCost)
-        InspectorInstrumentation::didChangeCanvasMemory(*context);
+    m_contextStateSaver = nullptr;
+    RefPtr returnBuffer = std::exchange(m_imageBuffer, WTFMove(buffer));
 
+    IntSize oldSize = m_size;
+    size_t oldMemoryCost = m_imageBufferMemoryCost.load(std::memory_order::relaxed);
+    size_t newMemoryCost = 0;
     if (m_imageBuffer) {
+        m_size = m_imageBuffer->truncatedLogicalSize();
+        newMemoryCost = m_imageBuffer->memoryCost();
         m_imageBuffer->context().setShadowsIgnoreTransforms(true);
         m_imageBuffer->context().setImageInterpolationQuality(defaultInterpolationQuality);
         m_imageBuffer->context().setStrokeThickness(1);
         m_contextStateSaver = makeUnique<GraphicsContextStateSaver>(m_imageBuffer->context());
-
+    }
```

Comment:
You don't change `m_size` if `m_imageBuffer` is null. Is that OK?

Reply:
Yeah, it's the existing behavior, did not want to change it.
There's some buffer stealing behavior with `setImageBuffer(nullptr)` that is not a resize operation. But future patches should remove those calls, they're a bit hacky. In that work the RefPtr here should be changed to Ref